cmd/process-user-data/update.go

```diff
@@ -94,14 +96,92 @@ func updateAgentConfig(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("failed to get daemon config from local file")
 	}
 
-	// Replace the value of aa_kbc_params in the agent config file - default: /etc/agent-config.toml
-	// with the value of aa-kbc-params
-	if err := updateAAKBCParams(config.AAKBCParams, cfg.agentConfigPath); err != nil {
-		fmt.Printf("Error: Failed to update agent config file with aa-kbc-params: %s\n", err)
-		return err
+	// Parse the agent config file
+	agentConfig, err := parseAgentConfig(cfg.agentConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to parse agent config file: %s", err)
+	}
+
+	if config.AAKBCParams != "" {
+		fmt.Printf("Updating aa_kbc_params in agent config file")
+		agentConfig.AaKbcParams = config.AAKBCParams
+	}
+
+	if config.AuthJson != "" {
+		// Check if AaKbcParams is offline_fs_kbc
+		if strings.Contains(agentConfig.AaKbcParams, "offline_fs_kbc") {
+			// Create offline registry auth file
+			// Create the file
+			file, err := os.Create(offlineKbcAuthFile)
+			if err != nil {
+				return fmt.Errorf("failed to create file: %s", err)
+			}
+			defer file.Close()
+
+			_, err = file.WriteString(cloudinit.AuthJSONToResourcesJSON(string(config.AuthJson)))
+			if err != nil {
+				return fmt.Errorf("failed to write agent config file: %s", err)
+			}
+
+		} else if strings.Contains(agentConfig.AaKbcParams, "cc_kbc") {
+
+			fmt.Printf("Updating image_registry_auth_file in agent config file with value\n")
+			// Check if authJsonFilePath exists. If it exists update the file path in the
+			// agent config
+			if _, err := os.Stat(authJsonFilePath); err == nil {
+				// Update the file path in the agent config
+				agentConfig.ImageRegistryAuthFile = "file://" + authJsonFilePath
+			} else {
+				fmt.Printf("auth.json file doesn't exist. Not updating the image_registry_auth_file in agent config file\n")
+			}
+		} else {
+			fmt.Printf("Unknown KBC option. Not updating the agent config file\n")
+		}
```

Comment:
this is quite nested. can we factor that into it's own function, .e.g `setAuthJsonConfig(agentConfig, config.AuthJson)`?

Reply:
Yeah, @stevenhorsman also suggested to set this for any kbc.